ONSdigital/eq-author-app
# EAR 574 - Add skip conditions to folders

### What is the context of this PR?

Allows users to add skip conditions to folders so that their entire contents can be skipped in Runner when rules are met.

[Ticket](https://collaborate2.ons.gov.uk/jira/browse/EAR-574)

### Changes
- Adds ability to add skip logic using Author UI for folders
- Adds a placeholder routing page for folders stating that routing rules are not supported (similar to question confirmation "no routing" page)
- Disables adding skip conditions to first folder in questionnaire
- GraphQL changes (minor) to add skip conditions to folders
- Adds `displayName` resolver to folders for consistency with other entities - returns alias or "Untitled folder" - rejig folder design page on UI to use this instead
- Publisher changes to enable output of skip conditions on folders - adds skip condition to the runner group
- Add support for validation errors on folders - skip condition validation errors should work as before & red dot displays next to tabs / folder name in nav when present

### How to review
- Add a folder
- Add skip conditions to a folder - check usual skip condition functionality works as before (e.g. answers must be from previous questions, etc.)
- Check validation errors on skip conditions work in the same way as for question pages / question confirmations
- Check that you can't add skip conditions to the first folder in a questionnaire
- Check appearance of folder's "no routing" page
- View survey & check skip conditions applied properly - should skip entire group of questions in folder if skip conditions met
- ☕️ 🫖

### Images
#### Routing placeholder page
<img width="1834" alt="Screenshot 2021-03-12 at 14 08 29" src="https://user-images.githubusercontent.com/60441612/110952331-003ada00-833e-11eb-8f10-90d52c248e9b.png">

#### Can't add skip conditions to first folder in questionnaire
<img width="1834" alt="Screenshot 2021-03-12 at 14 08 40" src="https://user-images.githubusercontent.com/60441612/110952346-04ff8e00-833e-11eb-98d8-fa69884a3b52.png">

#### Skip conditions
<img width="1834" alt="Screenshot 2021-03-12 at 14 19 30" src="https://user-images.githubusercontent.com/60441612/110952354-07fa7e80-833e-11eb-9470-81d2fceb2e66.png">


### What to do after everything is green

1. - [ ] Bring this branch up-to-date with Origin/Master
2. - [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. - [ ] Click the **Merge** button on this pull request
4. - [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. - [ ] Move the Jira ticket for this task into the next stage of the process
